### PR TITLE
Adding new query_key for elf flow and changing smi tests for elf flow

### PR DESCRIPTION
**Pull request by hlaccabu**
#### Problem solved by the commit
adding support for new elf flow for npu3 and other products. Change to two smi validate tests to demonstrate how the flow works.

#### What has been tested and how, request additional testing if necessary
Compiled and tested with npu3. Elf flow works with some additional changes to drivers.

### CODE REVIEW

File: src/runtime_src/core/tools/common/tests/TestNPULatency.cpp [hunk 1]
```diff
@@ -47,42 +50,75 @@ TestNPULatency::run(std::shared_ptr<xrt_core::device> dev)
...
   auto working_dev = xrt::device(dev);
   working_dev.register_xclbin(xclbin);
 
   xrt::hw_context hwctx;
   xrt::kernel testker;
-  try {
-    hwctx = xrt::hw_context(working_dev, xclbin.get_uuid());
-    testker = xrt::kernel(hwctx, kernelName);
+ 
+  if(xrt_core::device_query<xrt_core::query::rom_vbnv>(dev).find("npu3") == std::string::npos) {
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/tests/TestNPULatency.cpp:60]: Nit. consistency please :-)
```suggestion
  if (xrt_core::device_query<xrt_core::query::rom_vbnv>(dev).find("npu3") == std::string::npos) {
```

File: src/runtime_src/core/tools/common/tests/TestNPULatency.cpp [hunk 2]
```diff
@@ -47,42 +50,75 @@ TestNPULatency::run(std::shared_ptr<xrt_core::device> dev)
...
+      elf = xrt::elf(elf_path);
+    }
+    catch (const std::runtime_error& ex) {
+      logger(ptree, "Error", ex.what());
+      ptree.put("status", test_token_failed);
+      return ptree;
+    }
+
+    xrt::module mod{elf};
+    try{
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/tests/TestNPULatency.cpp:111]:
```suggestion
    try {
```

File: src/runtime_src/core/tools/common/tests/TestNPULatency.cpp [hunk 3]
```diff
@@ -47,42 +50,75 @@ TestNPULatency::run(std::shared_ptr<xrt_core::device> dev)
...
+    xrt::xclbin::kernel xkernel;
+    if (itr!=xkernels.end())
+      xkernel = *itr;
+    else {
+      logger(ptree, "Error", "No kernel with `DPU` found in the xclbin");
+      ptree.put("status", test_token_failed);
+      return ptree;
+    }
+    auto kernelName = xkernel.get_name();
+    if(XBU::getVerbose())
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/tests/TestNPULatency.cpp:78]:
```suggestion
    if (XBU::getVerbose())
```

**Review by stsoe — APPROVED** [review]
LGTM.  Some consistency suggestions

File: src/runtime_src/core/tools/common/TestRunner.cpp [hunk 4]
```diff
@@ -380,6 +380,29 @@ TestRunner::search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& d
...
+ * Function to run elf flow for xrt_smi tests
+*/
+xrt::device
+TestRunner::elf_flow(const std::shared_ptr<xrt_core::device>& dev, const xrt::xclbin& xclbin, xrt::hw_context& hwctx,
+              xrt::kernel& testker, boost::property_tree::ptree& ptTest)
+{
+  const auto elf_name = xrt_core::device_query<xrt_core::query::elf_name>(dev, xrt_core::query::elf_name::type::nop);
+  auto elf_path = findPlatformFile(elf_name, ptTest);
+
+  logger(ptTest, "Elf", elf_path);
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/TestRunner.cpp:393]: We are removing logging path names, please remove this

File: src/runtime_src/core/tools/common/TestRunner.cpp [hunk 5]
```diff
@@ -380,6 +380,29 @@ TestRunner::search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& d
...
+
+  auto working_dev = xrt::device(dev);
+  working_dev.register_xclbin(xclbin);
+  xrt::elf elf;
+  elf = xrt::elf(elf_path);
+  xrt::module mod{elf};
+
+  hwctx = xrt::hw_context{working_dev, xclbin.get_uuid()};
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/TestRunner.cpp:401]: xclbin shouldn't be here?

**Review by AShivangi — APPROVED** [review]
Looks good. Thanks!

File: src/runtime_src/core/tools/common/TestRunner.cpp [hunk 6]
```diff
@@ -380,6 +380,62 @@ TestRunner::search_and_program_xclbin(const std::shared_ptr<xrt_core::device>& d
...
+    xrt::xclbin::kernel xkernel;
+    if (itr!=xkernels.end())
+      xkernel = *itr;
+    else {
+      logger(ptTest, "Error", "No kernel with `DPU` found in the xclbin");
+      ptTest.put("status", test_token_failed);
+    }
+    auto kernelName = xkernel.get_name();
+    if(XBUtilities::getVerbose())
+      logger(ptTest, "Details", boost::str(boost::format("Kernel name is '%s'") % kernelName));
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/TestRunner.cpp:408]: Please remove this log

**Review by rchane — APPROVED** [review]

**Pull request merged**